redwood-treehouse-host/src/iosTest/kotlin/app/cash/redwood/treehouse/IosTreehouseDispatchersTest.kt

```diff
@@ -15,10 +15,29 @@
  */
 package app.cash.redwood.treehouse
 
+import app.cash.turbine.test
+import kotlin.test.Test
+import kotlinx.coroutines.Runnable
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.consumeAsFlow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.test.runTest
+
 class IosTreehouseDispatchersTest : AbstractTreehouseDispatchersTest() {
   override val treehouseDispatchers: TreehouseDispatchers = IosTreehouseDispatchers()
 
   /** We haven't set done the work to dispatch to the UI thread on iOS tests. */
   override val ignoreTestsThatExecuteOnUiThread: Boolean
     get() = true
+
+  @Test
+  fun closeOnlyStopsNewZiplineRunnable() = runTest {
+    val channel = Channel<Unit>()
+    val runnable = Runnable { launch { channel.send(Unit) } }
+    treehouseDispatchers.zipline.dispatch(coroutineContext, runnable)
+    treehouseDispatchers.close()
+    treehouseDispatchers.zipline.dispatch(coroutineContext, runnable)
+
+    channel.consumeAsFlow().test { awaitItem() }
```

Comment:
Unfortunately this isn't sufficient to test the absence of the second invocation of the `Runnable`. Turbine can only help you wait and test for the presence of events.

In general there's no way to test for absence without reaching into implementation details. I think the best we can do here is to make `_zipline` and `SingleThreadCoroutineDispatcher` be `internal` and then reach into that to get at the `NSThread` and wait (with a timeout) until it exits.

With that we don't even need the channel or to schedule work, since we can just call close and ensure the thread goes away. Of course we're writing a bit of a change-detector test exposing the internals, but it's still fairly high level in that we only ensure we don't leak the thread and aren't too privy to the internal mechanisms.

Reply:
> Unfortunately this isn't sufficient to test the absence of the second invocation of the Runnable. Turbine can only help you wait and test for the presence of events.

Ah, I thought this was working without an exception for not all items consumed being thrown.

> I think the best we can do here is to make _zipline and SingleThreadCoroutineDispatcher be internal and then reach into that to get at the NSThread and wait (with a timeout) until it exits.

Sounds good.

> Of course we're writing a bit of a change-detector test exposing the internals, but it's still fairly high level in that we only ensure we don't leak the thread and aren't too privy to the internal mechanisms.

Yeah, I was trying to avoid diving into the internals too much with the first approach, but this makes sense.